src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java

```diff
@@ -1054,14 +1058,41 @@ public boolean hasChildren(Resource resource) {
     public String getParentResourceType(final Resource resource) {
         String resourceSuperType = null;
         if ( resource != null ) {
-            resourceSuperType = resource.getResourceSuperType();
-            if (resourceSuperType == null) {
-                resourceSuperType = this.getParentResourceType(resource.getResourceType());
+            if (getParentResourceTypeMap().containsKey(resource.getPath())) {
+                resourceSuperType = getParentResourceTypeMap().get(resource.getPath());
+            } else {
+                resourceSuperType = getParentResourceTypeInternal(resource);
+                getParentResourceTypeMap().put(resource.getPath(), resourceSuperType);
```

Comment:
Can be simplified:

```suggestion
            resourceSuperType = getParentResourceTypeMap().computeIfAbsent(resource.getPath(), p -> getParentResourceTypeInternal(resource));
```

Reply:
I don't think that this will work, because it lead to a recursive call:
``getPrentResourceType`` -> ``getParentResourceTypeInternal`` -> ``getParentResourceType``

and that as part of ``computeIfAbsent``  multiple keys are modified in the map returned by ``getParentResourceTypeMap()``, which can lead to a ``ConcurrentModificationException``.